enhance: use tag for federating quotes

Johann150 opened the pull request:
# What
Use a `Link` tag to federate quotes instead of the custom `_misskey_quote` property.

# Why
related to #8722

# Additional info
This uses the same namespace URL that the `_misskey_quote` property uses in JSON-LD for the `rel` attribute.

With this change the deprecated properties will still be sent so this is for now fully backwards compatible. It should be reviewed at a later date whether the `_misskey_quote` property (and maybe the `quoteUrl` property as well) are still necessary or should be removed to reduce network traffic.

Johann150 committed:
enhance: add tag for quotes

Johann150 committed:
enhance: parse quote tag

ThatOneCalculator commented on packages/backend/src/remote/activitypub/renderer/note.ts:
```diff
@@ -111,6 +111,16 @@ export default async function renderNote(note: Note, dive = true, isTalk = false
+	if (quote) {
+		tag.push({
+			type: 'Link',
+			rel: 'https://misskey-hub.net/ns#_misskey_quote',
```
Seems to be a dead/invalid link? 
![image](https://user-images.githubusercontent.com/44733677/175430712-427bcf16-9800-458e-bf2e-9b5b0bb16484.png)

tamaina replied:
no problem

mei23 replied:
We are using URIs to make the namespace unique. So there is no problem if you can't access it.

Johann150 replied:
For completeness: This is as specified for ActivityPub vocabulary:
> ### `rel`
> A link relation associated with a [Link](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-link). The value MUST conform to both the [[HTML5](https://www.w3.org/TR/html5/)] and [[RFC5988](https://datatracker.ietf.org/doc/html/rfc5988#section-4)] "link relation" definitions.
-- https://www.w3.org/TR/activitystreams-vocabulary/#dfn-rel

The "link relation" definition from RFC 5988 allows two relation types, relevant is this one:

> ### 4.2. Extension Relation Types
> Applications [...] can use an extension relation type, which is a URI [...]. Although the URI can point to a resource that contains a definition of the semantics of the relation type, clients SHOULD NOT automatically access that resource [...].
-- https://datatracker.ietf.org/doc/html/rfc5988#section-4.2

Johann150 committed:
use attachment instead of tag

When federating out will use attachment instead of tag for compatibility
with existing software (zap, streams). On incoming content, both
attachment and tag will be scanned for fitting items.

mei23 commented:
attachment だけは意味的にありえないと思う。
というか他人の投稿を添付した扱いになるのは困る。

あと、`_misskey_quote` がダメな理由がわからない。
`quoteUrl` は ActivityStream を勝手に拡張しているかもしれないけど、`_misskey_quote` は一応 valid なはず。

noellabo commented:
うん、attachmentではないよね。

ちなみにFedibirdの引用はquoteUrlからquoteUriに変更してこうなってる。
```
@context: [{
  fedibird: "http://fedibird.com/ns#",
  quoteUri: "fedibird:quoteUri"
}],
quoteUri: "https://misskey.io/notes/91zs39f68n",
```

mei23 commented:
これらは全て valid であって 独自拡張 であるので、どれかに寄せる理由は何という感じなんだわ。
`"http://fedibird.com/ns#quoteUri": ...`
`"https://misskey-hub.net/ns#_misskey_quote": ...`
`tag rel="https://misskey-hub.net/ns#_misskey_quote"`

mei23 commented:
そもそも `quoteUri`, `_misskey_quote` は invalid という誤認を元に話を進めきているのが、変な感じ。
https://github.com/mastodon/mastodon/issues/18473#issuecomment-1133874921
↓
https://github.com/misskey-dev/misskey/issues/8722
↓

Johann150 commented:
@mei23 The discussion is taking place in the linked issue, why do you not participate there? Just complaining to me is not that helpful.

> そもそも quoteUri, _misskey_quote は invalid という誤認を元に話を進めきているのが、変な感じ。

... as I said please read the issue, that topic is covered there and you can comment on it there:

> I'm not sure why the current implementation seems to be completely out of the picture, seeing that it is seemingly starting to be adopted by other ActivityPub implementations (Fedibird, Pleroma, as I also laid out on the Mastodon issue).
-- https://github.com/misskey-dev/misskey/issues/8722#issuecomment-1134008835

Johann150 commented:
I am also not sure why you are complaining about that here because this pull request does not remove or modify anything related to `quoteUrl` and `_misskey_quote`.

mei23 commented:
議論はIssueの方でするけど、この Pull Request が Issueのどこで合意を経て `tag` にしたのか `attachment` にしたのかがわからない。

mei23 commented:
> I am also not sure why you are complaining about that here because this pull request does not remove or modify anything related to quoteUrl and _misskey_quote.

Because it was written here that it will be removed later
> the _misskey_quote property (and maybe the quoteUrl property as well) should be removed at a later time.

mei23 commented:
変える必要がなさそうで仕様も決まってないならこのプルリクは何をするものなの？
Draftでもないならもうこの仕様にしようとしていることじゃないの？
とかすごく納得がいかないわ

mei23 commented:
https://github.com/misskey-dev/misskey/pull/8837#issuecomment-1166894145

とりあえず、Misskeyの互換用のも変えておいたほうが良さそう
`https://www.w3.org/ns/activitystreams#quoteUrl`
↓
`http://fedibird.com/ns#quoteUri`

rinsuki commented:
そもそもquoteUrlはfedibirdじゃない別のMastodonのfork用だったはず (今でも生きてる鯖があるかは不明)

Johann150 commented:
> Because it was written here that it will be removed later
>> the _misskey_quote property (and maybe the quoteUrl property as well) should be removed at a later time.

I have updated the description.

Johann150 committed:
deduplicate Link by href

Johann150 committed:
Revert "use attachment instead of tag"

This partially reverts commit d4ad786d3e35b57af6839bd5102c563d4cecf05a,
specifically the part of the renderer. Quotes will now be rendered inside
tag.

Note that parsing will still check both the attachment and tag properties.

Johann150 commented:
I have added deduplication based on the `href` field, as this implementation parses both the `tag` and the `attachment` field and in the issue it was suggested that the object may be added to both.

I have also reverted back to using `tag` as it was suggested as the "technically correct" solution.

silverpill commented on packages/backend/src/remote/activitypub/models/tag.ts:
```diff
@@ -16,3 +16,34 @@ export function extractApHashtagObjects(tags: IObject | IObject[] | null | undef
+	} else {
+		// either multiple quotes (unsupported) or no quote
+		return null;
+	}
```
Other servers may allow multiple quotes per post. Perhaps it would be better to take the first link from the array if `quotes.length >= 1`?

trwnh commented on packages/backend/src/remote/activitypub/models/tag.ts:
```diff
@@ -16,3 +16,34 @@ export function extractApHashtagObjects(tags: IObject | IObject[] | null | undef
+
+	let quotes: ILink[] = toArray(tags)
+		.filter(isLink)
+		.filter(link => link.mediaType === 'application/activity+json');
```
Is this step necessary? `mediaType` isn't required for full semantic understanding. It seems adequate to check only for Links, which ideally have the specified `rel` that you check for immediately below this.

Consider the following JSON:

```json
"tag": [
  {
    "type": "Link",
    "href": "URI",
    "name": "RE: URI",
    "rel": "https://misskey-hub.net/ns#_misskey_quote"
  }
]
```

The current code would ignore this even though it is clearly intended as a misskey quote.

trwnh replied:
One other thing worth pointing out is that [on the associated FEP, I proposed](https://codeberg.org/fediverse/fep/pulls/13#issuecomment-601588) using `rel` instead of `mediaType` as a less ambiguous way of detecting when a Link refers to an ActivityStreams object. Something like this:

```json
"tag": [
  {
    "type": "Link",
    "href": "URI",
    "name": "RE: URI",
    "rel": ["https://www.w3.org/ns/activitystreams", "https://misskey-hub.net/ns#_misskey_quote"]
  }
]
```

Given that the FEP is for "Object Links"

trwnh replied:
Logically I would restructure the flow of this function along the following lines:

1) Check if there's a Link with link-rel of _misskey_quote. If yes, stop.
2) Check if there are any Links with link-rel of activitystreams. If yes, parse for `name` within `content`. If a match is found, add a quote of that object (and optionally replace `name` in `content` with a link targeting `href`).

trwnh replied:
Update on the FEP again: given that clients MUST specify `Accept: application/ld+json; profile="https://www.w3.org/ns/activitystreams"` it might make sense to use this `mediaType` instead of `application/activity+json`, since it is possible that an ActivityPub server might exist where `application/activity+json` is not supported (it is only a SHOULD and not a MUST).

Johann150 commented:
There are many merge conflicts and I do not have time or motivation to refactor to the new code layout.
